File: aoc15/ocaml/day_08/day_08.ml
let input = Input.char_lists "input" in

let rec memsize chars =
  match chars with
  | [] -> 0
  | '"' :: rs -> memsize rs
  | '\\' :: 'x' :: _ :: _ :: rs
  | '\\' :: '\\' :: rs
  | '\\' :: '"' :: rs
  | _ :: rs ->
      1 + memsize rs
in

let rec encode chars =
  match chars with
  | [] -> []
  | x :: rs when x = '\\' || x = '"' -> '\\' :: x :: encode rs
  | x :: rs -> x :: encode rs
in
let encoded_len chars = List.length (encode chars) + 2 in

let mem = List.fold_left ( + ) 0 (List.map memsize input) in
let code = List.fold_left ( + ) 0 (List.map List.length input) in
let encoded = List.fold_left ( + ) 0 (List.map encoded_len input) in
let first = code - mem in
let second = encoded - code in
Printf.printf "First: %d\n" first;
Printf.printf "Second: %d\n" second

File: aoc15/go/day_08/day_08.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func memsize(line string) (res int) {
	for i := 1; i < len(line)-1; i++ {
		if line[i] == '\\' && line[i+1] == 'x' {
			i += 3
		} else if line[i] == '\\' {
			i++
		}
		res++
	}
	return
}

func encoded(line string) (res int) {
	for i := 0; i < len(line); i++ {
		if line[i] == '"' || line[i] == '\\' {
			res++
		}
		res++
	}
	res += 2
	return
}

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)

	first := 0
	second := 0
	for scanner.Scan() {
		line := scanner.Text()
		first += len(line) - memsize(line)
		second += encoded(line) - len(line)
	}
	fmt.Println("First:", first)
	fmt.Println("Second:", second)
}
